### PR TITLE
[Refactor] Force reorder columns of chunk

### PR DESCRIPTION
**Pull request by DorianZheng**
Signed-off-by: dorianzheng <xingzhengde72@gmail.com>

## What type of PR is this：
- [ ] bug
- [ ] feature
- [ ] enhancement
- [ ] refactor
- [x] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

The column order of chunk is required to be invariable. In order to simplify the logic of each scanner, we force to reorder the columns of chunk, so scanner doesn't have to care about the column order anymore. The overhead of reorder is negligible because we don't actually copy columns data.

### CODE REVIEW

**Comment by DorianZheng** [comment]
I will add ut later

**Review by dirtysalt — DISMISSED** [review]

File: be/src/storage/chunk_helper.cpp [hunk 1]
```diff
@@ -373,4 +373,19 @@ std::shared_ptr<vectorized::Chunk> ChunkHelper::new_chunk(const std::vector<Slot
...
 
+void ChunkHelper::reorder_chunk(const TupleDescriptor& tuple_desc, vectorized::Chunk* chunk) {
+    return reorder_chunk(tuple_desc.slots(), chunk);
+}
+
+void ChunkHelper::reorder_chunk(const std::vector<SlotDescriptor*>& slots, vectorized::Chunk* chunk) {
+    DCHECK(chunk->columns().size() == slots.size());
+    auto reordered_chunk = vectorized::Chunk();
+    auto& original_chunk = (*chunk);
+    for (std::size_t idx = 0; idx < slots.size(); idx++) {
```
Comment (imay) [inline be/src/storage/chunk_helper.cpp:384]: IMO, not all slots will be read in chunk. So there will be slot in `slots` which has no corresponding column in chunk.
So I think you should continue rather than push a nullable in to reordered_chunk

Comment (DorianZheng) [reply be/src/storage/chunk_helper.cpp:384]: I have talked with @kangkaisen. He said if there is slot, there will be corresponding column

**Comment by DorianZheng** [comment]
waiting for daily case passed: http://39.101.206.100:8090/view/Daily/job/master_asan/600/

**Review by imay — DISMISSED** [review]

**Comment by DorianZheng** [comment]
rebase and run daily case again: http://39.101.206.100:8090/view/Daily/job/master_asan/607/

**Comment by DorianZheng** [comment]
> rebase and run daily case again: http://39.101.206.100:8090/view/Daily/job/master_asan/607/

I aborted it due to previous run haved crashed. I will run it again when it's fixed.

**Comment by DorianZheng** [comment]
http://39.101.206.100:8090/view/regression/job/starrocks_pr_regression/576/#showFailuresLink

I've checked the newest tests, they are not caused by this PR

**Review by dirtysalt — APPROVED** [review]

**Review by imay — APPROVED** [review]

**Comment by DorianZheng** [comment]
image

File: be/src/storage/chunk_helper.cpp [hunk 2]
```diff
@@ -373,4 +373,19 @@ std::shared_ptr<vectorized::Chunk> ChunkHelper::new_chunk(const std::vector<Slot
     return chunk;
 }
 
+void ChunkHelper::reorder_chunk(const TupleDescriptor& tuple_desc, vectorized::Chunk* chunk) {
+    return reorder_chunk(tuple_desc.slots(), chunk);
+}
+
+void ChunkHelper::reorder_chunk(const std::vector<SlotDescriptor*>& slots, vectorized::Chunk* chunk) {
```
Comment (dirtysalt) [inline be/src/storage/chunk_helper.cpp:380]: I think we can do some optimization: that if order is correct, we don't need to do anything at all. 

But we can put this opt into another PR.

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: be/src/storage/chunk_helper.cpp (hunk 2 above)
Comment (DorianZheng) [reply be/src/storage/chunk_helper.cpp:380]: If so, I have to for-loop twice, first is to check if the order is correct, second is to reorder it if the order is not correct. I'm not sure the benefit of not swapping columns deserves doing this.

Comment (dirtysalt) [reply be/src/storage/chunk_helper.cpp:380]: for-loop is just CPU operation, faster than allocate memory for chunk. And you can make for-loop fast with some efforts.

Comment (DorianZheng) [reply be/src/storage/chunk_helper.cpp:380]: There is few memory allocation for chunk, chunk only holds shared pointer to columns. The typical memory footprint of a `std::shared_ptr` is 44B. Say there are 1k columns in the chunk, the memory footprint of chunk will be ~44KB

Comment (dirtysalt) [reply be/src/storage/chunk_helper.cpp:380]: It's not about memory footprint, it's about memory allocation.

Allocate 1k times is bad. and think in chunk there are map & vector data structures.

Comment (DorianZheng) [reply be/src/storage/chunk_helper.cpp:380]: The memory allocation is not linear, typical memory allocation strategy is twice the space. So assume there are 1k columns, it will only allocate 10 times for `Columns`

Comment (dirtysalt) [reply be/src/storage/chunk_helper.cpp:380]: OK.
